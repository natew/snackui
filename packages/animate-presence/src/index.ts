export * from './AnimatePresence'
export * from './usePresence'
export { PresenceContext } from './PresenceContext'
