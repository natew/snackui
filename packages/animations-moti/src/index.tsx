import './polyfill'

export * from './create-animations'
