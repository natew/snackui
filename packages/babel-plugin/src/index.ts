import { basename } from 'path'

import generator from '@babel/generator'
import { declare } from '@babel/helper-plugin-utils'
import template from '@babel/template'
import { Visitor } from '@babel/traverse'
import * as t from '@babel/types'
import { simpleHash } from '@tamagui/simple-hash'
import {
  createExtractor,
  getPragmaOptions,
  isSimpleSpread,
  literalToAst,
} from '@tamagui/static'
import type { TamaguiOptions } from '@tamagui/static'

const importNativeView = template(`
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
`)

const importStyleSheet = template(`
const __ReactNativeStyleSheet = require('react-native').StyleSheet;
`)

const importWithTheme = template(`
const __internalWithTheme = require('@tamagui/core').internalWithTheme;
`)

const extractor = createExtractor()

export default declare(function snackBabelPlugin(
  api,
  options: TamaguiOptions
): {
  name: string
  visitor: Visitor
} {
  api.assertVersion(7)

  return {
    name: 'tamagui',

    visitor: {
      Program: {
        enter(this: any, root) {
          let sourcePath = this.file.opts.filename

          // this filename comes back incorrect in react-native, it adds /ios/ for some reason
          // adding a fix here, but it's a bit tentative...
          if (process.env.SOURCE_ROOT?.endsWith('ios')) {
            sourcePath = sourcePath.replace('/ios', '')
          }

          let hasImportedView = false
          let hasImportedViewWrapper = false
          const sheetStyles = {}
          const sheetIdentifier = root.scope.generateUidIdentifier('sheet')
          const firstComment =
            root.node.body[0]?.leadingComments?.[0]?.value?.trim() ?? ''
          const { shouldPrintDebug, shouldDisable } = getPragmaOptions({
            disableCommentCheck: true,
            source: firstComment,
            path: sourcePath,
          })

          if (shouldDisable) {
            return
          }

          function addSheetStyle(style: any, node: t.JSXOpeningElement) {
            const styleIndex = `${Object.keys(sheetStyles).length}`
            let key = `${styleIndex}`
            if (process.env.NODE_ENV === 'development') {
              const lineNumbers = node.loc
                ? node.loc.start.line +
                  (node.loc.start.line !== node.loc.end.line
                    ? `-${node.loc.end.line}`
                    : '')
                : ''
              key += `:${basename(sourcePath)}:${lineNumbers}`
            }
            sheetStyles[key] = style
            return readStyleExpr(key)
          }

          function readStyleExpr(key: string) {
            return template(`SHEET['KEY']`)({
              SHEET: sheetIdentifier.name,
              KEY: key,
            })['expression'] as t.MemberExpression
          }

          try {
            extractor.parseSync(root, {
              // @ts-expect-error in case they leave it out
              platform: 'native',
              shouldPrintDebug,
              importsWhitelist: ['constants.js', 'colors.js'],
              deoptProps: new Set([
                'focusStyle',
                'hoverStyle',
                'pressStyle',
                'pointerEvents',
              ]),
              extractStyledDefinitions: options.forceExtractStyleDefinitions,
              excludeProps: new Set([
                'className',
                'userSelect',
                'whiteSpace',
                'textOverflow',
                'cursor',
                'contain',
              ]),
              ...options,
              // disable this extraction for now at least, need to figure out merging theme vs non-theme
              // because theme need to stay in render(), whereas non-theme can be extracted
              // for now just turn it off entirely at a small perf loss
              disableExtractInlineMedia: true,
              // disable extracting variables as no native concept of them (only theme values)
              disableExtractVariables: options.experimentalFlattenThemesOnNative
                ? false
                : 'theme',
              sourcePath,

              // disabling flattening for now
              // it's flattening a plain <Paragraph>hello</Paragraph> which breaks things because themes
              // thinking it's not really worth the effort to do much compilation on native
              // for now just disable flatten as it can only run in narrow places on native
              // disableFlattening: 'styled',

              getFlattenedNode({ isTextView }) {
                if (!hasImportedView) {
                  hasImportedView = true
                  root.unshiftContainer('body', importNativeView())
                }
                return isTextView ? '__ReactNativeText' : '__ReactNativeView'
              },

              onExtractTag(props) {
                const { isFlattened } = props

                if (!isFlattened) {
                  // we aren't optimizing at all if not flattened anymore
                  return
                }

                assertValidTag(props.node)
                const stylesExpr = t.arrayExpression([])
                const finalAttrs: (t.JSXAttribute | t.JSXSpreadAttribute)[] = []
                const themeKeysUsed = new Set<string>()

                function getStyleExpression(style: Object | null) {
                  if (!style) return

                  // split theme properties and leave them as props since RN has no concept of theme
                  const { plain, themed } = splitThemeStyles(style)

                  const ident = addSheetStyle(plain, props.node)

                  if (themed && options.experimentalFlattenThemesOnNative) {
                    for (const key in themed) {
                      themeKeysUsed.add(key)
                    }

                    // make a sub-array
                    return t.arrayExpression([ident, addThemedStyleExpression(themed)])
                  } else {
                    // since we only do flattened disabling this path
                    return ident
                    // when we supported extracting non-flattened
                    // addStyleExpression(ident, isFlattened ? simpleHash(JSON.stringify(plain)) : undefined)
                  }
                }

                function addStyleExpression(expr: any) {
                  stylesExpr.elements.push(expr)
                }

                function addThemedStyleExpression(styles: Object) {
                  const themedStylesAst = literalToAst(styles) as t.ObjectExpression
                  themedStylesAst.properties.forEach((_) => {
                    const prop = _ as t.ObjectProperty
                    if (prop.value.type === 'StringLiteral') {
                      prop.value = t.memberExpression(
                        t.identifier('theme'),
                        t.identifier(prop.value.value.slice(1))
                      )
                    }
                  })
                  return themedStylesAst
                }

                for (const attr of props.attrs) {
                  switch (attr.type) {
                    case 'style': {
                      addStyleExpression(getStyleExpression(attr.value))
                      break
                    }
                    case 'ternary': {
                      const { consequent, alternate } = attr.value

                      const consExpr = getStyleExpression(consequent)
                      const altExpr = getStyleExpression(alternate)

                      const styleExpr = t.conditionalExpression(
                        attr.value.test,
                        consExpr || t.nullLiteral(),
                        altExpr || t.nullLiteral()
                      )
                      addStyleExpression(
                        styleExpr
                        // isFlattened ? simpleHash(JSON.stringify({ consequent, alternate })) : undefined
                      )
                      break
                    }
                    case 'attr': {
                      if (t.isJSXSpreadAttribute(attr.value)) {
                        if (isSimpleSpread(attr.value)) {
                          stylesExpr.elements.push(
                            t.memberExpression(attr.value.argument, t.identifier('style'))
                          )
                        }
                      }
                      finalAttrs.push(attr.value)
                      break
                    }
                  }
                }

                props.node.attributes = finalAttrs

                if (props.isFlattened) {
                  if (themeKeysUsed.size) {
                    if (!hasImportedViewWrapper) {
                      root.unshiftContainer('body', importWithTheme())
                      hasImportedViewWrapper = true
                    }

                    const name = props.node.name['name']
                    const WrapperIdentifier = root.scope.generateUidIdentifier(
                      name + 'Wrapper'
                    )

                    root.pushContainer(
                      'body',
                      t.variableDeclaration('const', [
                        t.variableDeclarator(
                          WrapperIdentifier,
                          t.callExpression(t.identifier('__internalWithTheme'), [
                            t.identifier(name),
                            t.arrowFunctionExpression(
                              [t.identifier('theme')],
                              t.blockStatement([t.returnStatement(stylesExpr)])
                            ),
                            t.arrayExpression(
                              [...themeKeysUsed].map((k) => t.stringLiteral(k))
                            ),
                          ])
                        ),
                      ])
                    )

                    // @ts-ignore
                    props.node.name = WrapperIdentifier
                  } else {
                    props.node.attributes.push(
                      t.jsxAttribute(
                        t.jsxIdentifier('style'),
                        t.jsxExpressionContainer(
                          stylesExpr.elements.length === 1
                            ? (stylesExpr.elements[0] as any)
                            : stylesExpr
                        )
                      )
                    )
                  }
                }
              },
            })
          } catch (err) {
            if (err instanceof Error) {
              // metro doesn't show stack so we can
              let message = `${shouldPrintDebug === 'verbose' ? err : err.message}`
              if (message.includes('Unexpected return value from visitor method')) {
                message = 'Unexpected return value from visitor method'
              }
              console.warn('Error in Tamagui parse, skipping', message, err.stack)
              return
            }
          }

          if (!Object.keys(sheetStyles).length) {
            if (shouldPrintDebug) {
              console.info('END no styles')
            }
            return
          }

          const sheetObject = literalToAst(sheetStyles)
          const sheetOuter = template(
            'const SHEET = __ReactNativeStyleSheet.create(null)'
          )({
            SHEET: sheetIdentifier.name,
          }) as any

          // replace the null with our object
          sheetOuter.declarations[0].init.arguments[0] = sheetObject
          root.unshiftContainer('body', sheetOuter)
          // add import
          root.unshiftContainer('body', importStyleSheet())

          if (shouldPrintDebug) {
            console.info('\n -------- output code ------- \n')
            console.info(
              generator(root.parent)
                .code.split('\n')
                .filter((x) => !x.startsWith('//'))
                .join('\n')
            )
          }
        },
      },
    },
  }
})

function assertValidTag(node: t.JSXOpeningElement) {
  if (node.attributes.find((x) => x.type === 'JSXAttribute' && x.name.name === 'style')) {
    // we can just deopt here instead and log warning
    // need to make onExtractTag have a special catch error or similar
    if (process.env.DEBUG?.startsWith('tamagui')) {
      console.warn('⚠️ Cannot pass style attribute to extracted style')
    }
  }
}

function splitThemeStyles(style: Object) {
  const themed: Object = {}
  const plain: Object = {}
  let noTheme = true
  for (const key in style) {
    const val = style[key]
    if (val && val[0] === '$') {
      themed[key] = val
      noTheme = false
    } else {
      plain[key] = val
    }
  }
  return { themed: noTheme ? null : themed, plain }
}
