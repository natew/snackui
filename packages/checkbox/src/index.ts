import { CheckboxFrame, CheckboxIndicatorFrame } from './Checkbox'
import { createCheckbox } from './createCheckbox'

export * from './createCheckbox'
export * from './Checkbox'
export * from './CheckboxStyledContext'

export const Checkbox = createCheckbox({
  Frame: CheckboxFrame,
  Indicator: CheckboxIndicatorFrame,
})
