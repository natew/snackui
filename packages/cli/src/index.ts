#!/usr/bin/env node

process.on('beforeExit', () => {
  process.stdout.write(`exiting22123??????`)
})

process.on('SIGINT', () => {
  process.stdout.write(`exiting33123??????`)
})

import './cli.js'
