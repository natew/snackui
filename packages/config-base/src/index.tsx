export * from './tamagui.config'
export * from './media'
export * from './createGenericFont'
