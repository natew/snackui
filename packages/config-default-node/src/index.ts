// SEE config-default

import { CreateTamaguiProps, createFont, createTokens } from '@tamagui/core-node'
import { shorthands } from '@tamagui/shorthands'

// basic fallback theme just to have compiler load in decent tate
export function getDefaultTamaguiConfig() {
  const font = createFont({
    family: 'System',
    size: {
      1: 15,
    },
    lineHeight: {
      1: 15,
    },
    transform: {},
    weight: {
      1: '400',
    },
    color: {
      1: '$color',
    },
    letterSpacing: {
      1: 0,
    },
  })
  const size = {
    $0: 0,
    '$0.25': 2,
    '$0.5': 4,
    '$0.75': 8,
    $1: 20,
    '$1.5': 24,
    $2: 28,
    '$2.5': 32,
    $3: 36,
    '$3.5': 40,
    $4: 44,
    $true: 44,
    '$4.5': 48,
    $5: 52,
    $6: 64,
    $7: 74,
    $8: 84,
    $9: 94,
    $10: 104,
    $11: 124,
    $12: 144,
    $13: 164,
    $14: 184,
    $15: 204,
    $16: 224,
    $17: 224,
    $18: 244,
    $19: 264,
    $20: 284,
  }

  type SizeKeysIn = keyof typeof size
  type Sizes = {
    [Key in SizeKeysIn extends `$${infer Key}` ? Key : SizeKeysIn]: number
  }
  type SizeKeys = `${keyof Sizes extends `${infer K}` ? K : never}`

  const spaces = Object.entries(size).map(([k, v]) => {
    return [k, sizeToSpace(v)] as const
  })

  const spacesNegative = spaces.slice(1).map(([k, v]) => [`$-${k.slice(1)}`, -v])

  const space: {
    [Key in `$-{SizeKeys}` | SizeKeys]: Key extends keyof Sizes ? Sizes[Key] : number
  } = {
    ...Object.fromEntries(spaces),
    ...Object.fromEntries(spacesNegative),
  } as any

  // a bit odd but keeping backward compat for values >8 while fixing below
  function sizeToSpace(v: number) {
    if (v === 0) return 0
    if (v <= 8) return v * Math.log(v) * 0.12
    if (v <= 16) return Math.round(v * 0.333)
    return Math.floor(v * 0.7 - 12)
  }

  const zIndex = {
    0: 0,
    1: 100,
    2: 200,
    3: 300,
    4: 400,
    5: 500,
  }

  const radius = {
    0: 0,
    1: 3,
    2: 5,
    3: 7,
    4: 9,
    5: 10,
    6: 16,
    7: 19,
    8: 22,
    9: 26,
    10: 34,
    11: 42,
    12: 50,
  }

  const tokens = createTokens({
    color: {
      white: '#fff',
      black: '#000',
    },
    radius,
    zIndex,
    space,
    size,
  })

  const themes = {
    light: {
      background: tokens.color.white,
      color: tokens.color.black,
    },
    dark: {
      background: tokens.color.black,
      color: tokens.color.white,
    },
    // most of these used for testing:
    dark_blue: {
      background: 'blue',
      color: 'white',
    },
    dark_Card: {
      background: 'dark',
      color: 'card',
    },
    dark_blue_Card: {
      background: 'blue',
      color: 'white',
    },
    dark_blue_Button: {
      background: 'blue',
      color: 'white',
    },
    dark_red: {
      background: 'darkred',
      color: 'white',
    },
    dark_red_alt1: {
      background: 'darkred',
      color: 'white',
    },
    dark_red_alt1_ListItem: {
      background: 'red',
      color: 'red',
    },
    dark_red_active_ListItem: {
      background: 'darkred',
      color: 'red',
    },
    dark_red_Button: {
      background: 'darkred',
      color: '#ccc',
    },
    dark_yellow_Button: {
      background: 'brown',
      color: '#ccc',
    },
    dark_red_alt2: {
      background: 'darkred',
      color: '#555',
    },
    dark_red_alt2_Button: {
      background: 'darkred',
      color: '#444',
    },
    red: {
      background: 'red',
      color: 'red',
    },
  }

  return {
    shouldAddPrefersColorThemes: true,
    themeClassNameOnRoot: true,
    shorthands,
    fonts: {
      heading: font,
      body: font,
    },
    themes,
    tokens,
    media: {
      xs: { maxWidth: 660 },
      sm: { maxWidth: 800 },
      md: { maxWidth: 1020 },
      lg: { maxWidth: 1280 },
      xl: { maxWidth: 1420 },
      xxl: { maxWidth: 1600 },
      gtXs: { minWidth: 660 + 1 },
      gtSm: { minWidth: 800 + 1 },
      gtMd: { minWidth: 1020 + 1 },
      gtLg: { minWidth: 1280 + 1 },
      short: { maxHeight: 820 },
      tall: { minHeight: 820 },
      hoverNone: { hover: 'none' },
      pointerCoarse: { pointer: 'coarse' },
    },
  } satisfies CreateTamaguiProps
}
