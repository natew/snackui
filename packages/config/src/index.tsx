import { animations } from './animations'
import { configWithoutAnimations } from './config'
export * from './media'
export * from './createGenericFont'
export * from './animations'

export const config = {
  ...configWithoutAnimations,
  animations,
}
