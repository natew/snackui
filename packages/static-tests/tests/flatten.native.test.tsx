import * as React from 'react'
import { describe, expect, test } from 'vitest'

import { extractForNative } from './lib/extract'

Error.stackTraceLimit = Infinity
process.env.TAMAGUI_TARGET = 'native'
process.env.IS_STATIC = ''

window['React'] = React

describe('flatten-tests', () => {
  test('flattened media queries', async () => {
    const output = await extractForNative(`
      import { YStack } from 'tamagui/src/YStack'
      import { useMedia } from 'tamagui'
  
      export function Test(isLoading) {
        const media = useMedia()
        
        return (
          <YStack
            y={10}
            x={20}
            rotate="10deg"
            {...media.sm && {
              scale: 2,
              borderRadius: 10,
              backgroundColor: isLoading ? 'red' : 'blue'
            }}
          />
        )
      }
    `)
    const code = output?.code ?? ''

    expect(code).toMatchSnapshot()

    const startStr = `ReactNativeStyleSheet.create(`
    const start = code.indexOf(startStr)
    const end = code.indexOf(`});`)
    const defs = code.slice(start + startStr.length, end + 1)
    const sheetStyles = JSON.parse(defs)

    expect(sheetStyles['1']).toEqual({
      transform: [
        {
          translateY: 10,
        },
        {
          translateX: 20,
        },
        {
          rotate: '10deg',
        },
      ],
      flexDirection: 'column',
      flexShrink: 0,
      alignItems: 'stretch',
      minHeight: 0,
      minWidth: 0,
      position: 'relative',
      display: 'flex',
      flexBasis: 'auto',
    })

    expect(sheetStyles['2']).toEqual({
      borderTopLeftRadius: 10,
      borderTopRightRadius: 10,
      borderBottomRightRadius: 10,
      borderBottomLeftRadius: 10,
      transform: [
        {
          scale: 2,
        },
      ],
    })

    expect(sheetStyles['4']).toEqual({
      backgroundColor: 'rgba(255,0,0,1.00)',
    })

    expect(sheetStyles['6']).toEqual({
      backgroundColor: 'rgba(0,0,255,1.00)',
    })
  })
})
