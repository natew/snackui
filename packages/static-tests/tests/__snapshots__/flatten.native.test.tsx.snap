// Vitest Snapshot v1

exports[`flatten-tests > flattened media queries 1`] = `
"import { StyleSheet as ReactNativeStyleSheet } from 'react-native';
const _sheet = ReactNativeStyleSheet.create({
  \\"0\\": {},
  \\"1\\": {
    \\"alignItems\\": \\"stretch\\",
    \\"flexShrink\\": 0,
    \\"display\\": \\"flex\\",
    \\"flexBasis\\": \\"auto\\",
    \\"position\\": \\"relative\\",
    \\"minHeight\\": 0,
    \\"minWidth\\": 0,
    \\"flexDirection\\": \\"column\\",
    \\"transform\\": [{
      \\"translateY\\": 10
    }, {
      \\"translateX\\": 20
    }, {
      \\"rotate\\": \\"10deg\\"
    }]
  },
  \\"2\\": {
    \\"borderTopLeftRadius\\": 10,
    \\"borderTopRightRadius\\": 10,
    \\"borderBottomRightRadius\\": 10,
    \\"borderBottomLeftRadius\\": 10,
    \\"transform\\": [{
      \\"scale\\": 2
    }]
  },
  \\"3\\": {},
  \\"4\\": {
    \\"backgroundColor\\": \\"rgba(255,0,0,1.00)\\"
  },
  \\"5\\": {},
  \\"6\\": {
    \\"backgroundColor\\": \\"rgba(0,0,255,1.00)\\"
  },
  \\"7\\": {}
});
import { View as __ReactNativeView, Text as __ReactNativeText } from 'react-native';
import { YStack } from 'tamagui/src/YStack';
import { useMedia } from 'tamagui';
export function Test(isLoading) {
  const media = useMedia();
  return <__ReactNativeView style={[_sheet[\\"0\\"], _sheet[\\"1\\"], media.sm ? _sheet[\\"2\\"] : _sheet[\\"3\\"], media.sm && isLoading ? _sheet[\\"4\\"] : _sheet[\\"5\\"], media.sm && !isLoading ? _sheet[\\"6\\"] : _sheet[\\"7\\"]]} />;
}"
`;
